# SF-DAT-21 | Unit Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in Unit Projects 1 and 2.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.api as sm
import pylab as pl
%matplotlib inline

In [2]:
df_raw = pd.read_csv("../../dataset/admissions.csv")
df = df_raw.dropna()
print df.head()

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### Question 1. Let's create a frequency table of our variables.

In [4]:
df.prestige.unique()

array([ 3.,  1.,  4.,  2.])

In [5]:
df.prestige.count()

397

In [6]:
print np.nan, df.prestige.isnull().sum()
for prestige_score_count in np.sort(df.prestige.dropna().unique()):
    print prestige_score_count, len(df[df.prestige == prestige_score_count])

nan 0
1.0 61
2.0 148
3.0 121
4.0 67


In [7]:

admitted_prestige = pd.crosstab(index= df["admit"],  
                              columns=df["prestige"])      

admitted_prestige


prestige  1.0  2.0  3.0  4.0
admit                       
0          28   95   93   55
1          33   53   28   12

In [8]:

admitted_prestige = pd.crosstab(index= df["admit"],  
                              columns=df["gre"])      

admitted_prestige

gre    220.0  300.0  340.0  360.0  380.0  400.0  420.0  440.0  460.0  480.0  \
admit                                                                         
0          1      2      3      4      8      9      7      8      9     10   
1          0      1      1      0      0      2      0      2      4      6   

gre    ...    620.0  640.0  660.0  680.0  700.0  720.0  740.0  760.0  780.0  \
admit  ...                                                                    
0      ...       18     16     12     11     17      8      7      1      1   
1      ...       12      5     11      9      5      3      4      4      4   

gre    800.0  
admit         
0         14  
1         11  

[2 rows x 26 columns]

In [9]:

admitted_prestige = pd.crosstab(index= df["admit"],  
                              columns=df["gpa"])      

admitted_prestige

gpa    2.26  2.42  2.48  2.52  2.55  2.56  2.62  2.63  2.65  2.67  ...   3.90  \
admit                                                              ...          
0         1     1     1     1     1     1     1     1     0     1  ...      2   
1         0     1     0     0     0     0     1     0     1     1  ...      1   

gpa    3.91  3.92  3.93  3.94  3.95  3.97  3.98  3.99  4.00  
admit                                                        
0         1     2     1     4     1     1     0     2    15  
1         0     0     0     1     4     0     1     1    13  

[2 rows x 131 columns]

## Part 2. Return of dummy variables

#### Question 2.1. Create class or dummy variables for prestige.

In [12]:
dummy_ranks = pd.get_dummies(df.prestige, prefix = 'prestige')

In [13]:
dummy_ranks.rename(columns={'prestige_1.0': 'prestige_1',
                        'prestige_2.0': 'prestige_2',
                        'prestige_3.0': 'prestige_3',
                        'prestige_4.0': 'prestige_4'}, inplace = True)
dummy_ranks

prestige_1  prestige_2  prestige_3  prestige_4
0           0.0         0.0         1.0         0.0
1           0.0         0.0         1.0         0.0
2           1.0         0.0         0.0         0.0
3           0.0         0.0         0.0         1.0
4           0.0         0.0         0.0         1.0
5           0.0         1.0         0.0         0.0
6           1.0         0.0         0.0         0.0
7           0.0         1.0         0.0         0.0
8           0.0         0.0         1.0         0.0
9           0.0         1.0         0.0         0.0
10          0.0         0.0         0.0         1.0
11          1.0         0.0         0.0         0.0
12          1.0         0.0         0.0         0.0
13          0.0         1.0         0.0         0.0
14          1.0         0.0         0.0         0.0
15          0.0         0.0         1.0         0.0
16          0.0         0.0         0.0         1.0
17          0.0         0.0         1.0         0.0
18          0.0         1.0         0.0         0.0
19          1.0         0.0         0.0         0.0
20          0.0         0.0         1.0         0.0
21          0.0         1.0         0.0         0.0
22          0.0         0.0         0.0         1.0
23          0.0         0.0         0.0         1.0
24          0.0         1.0         0.0         0.0
25          1.0         0.0         0.0         0.0
26          1.0         0.0         0.0         0.0
27          0.0         0.0         0.0         1.0
28          0.0         1.0         0.0         0.0
29          1.0         0.0         0.0         0.0
..          ...         ...         ...         ...
370         0.0         1.0         0.0         0.0
371         0.0         0.0         1.0         0.0
372         1.0         0.0         0.0         0.0
373         1.0         0.0         0.0         0.0
374         0.0         1.0         0.0         0.0
375         0.0         0.0         0.0         1.0
376         0.0         1.0         0.0         0.0
377         0.0         1.0         0.0         0.0
378         0.0         0.0         1.0         0.0
379         0.0         1.0         0.0         0.0
380         0.0         1.0         0.0         0.0
381         0.0         1.0         0.0         0.0
382         0.0         1.0         0.0         0.0
383         1.0         0.0         0.0         0.0
384         0.0         1.0         0.0         0.0
385         1.0         0.0         0.0         0.0
386         0.0         1.0         0.0         0.0
387         0.0         1.0         0.0         0.0
388         0.0         1.0         0.0         0.0
389         0.0         1.0         0.0         0.0
390         0.0         1.0         0.0         0.0
391         0.0         1.0         0.0         0.0
392         0.0         0.0         1.0         0.0
393         0.0         1.0         0.0         0.0
394         0.0         0.0         1.0         0.0
395         0.0         1.0         0.0         0.0
396         0.0         0.0         1.0         0.0
397         0.0         1.0         0.0         0.0
398         0.0         1.0         0.0         0.0
399         0.0         0.0         1.0         0.0

[397 rows x 4 columns]

In [14]:
df = df.join([dummy_ranks])

In [15]:
df.columns


Index([u'admit', u'gre', u'gpa', u'prestige', u'prestige_1', u'prestige_2',
       u'prestige_3', u'prestige_4'],
      dtype='object')

In [16]:
df.drop('prestige', axis = 1, inplace = True)

In [17]:
df.columns

Index([u'admit', u'gre', u'gpa', u'prestige_1', u'prestige_2', u'prestige_3',
       u'prestige_4'],
      dtype='object')

#### Question 2.2. When modeling our class variables, how many do we need?

Answer: 3 classes are needed. 1 is needed for reference

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

Answer: # of variables - 1 (1 serves as a base). For this one in particular, we need 3 

In [18]:
cols_to_keep = ['admit', 'gre', 'gpa']
handCalc = df[cols_to_keep].join(prestige_df.ix[:, 'prestige_1':])
print handCalc.head()

   admit    gre   gpa  prestige_1.0  prestige_2.0  prestige_3.0  prestige_4.0
0      0  380.0  3.61           0.0           0.0           1.0           0.0
1      1  660.0  3.67           0.0           0.0           1.0           0.0
2      1  800.0  4.00           1.0           0.0           0.0           0.0
3      1  640.0  3.19           0.0           0.0           0.0           1.0
4      0  520.0  2.93           0.0           0.0           0.0           1.0


In [19]:

def percConvert(ser):
  return ser/float(ser[-1])

pd.crosstab(df.prestige_1, df.admit, margins=True).apply(percConvert, axis=1)

admit              0         1  All
prestige_1                         
0.0         0.723214  0.276786  1.0
1.0         0.459016  0.540984  1.0
All         0.682620  0.317380  1.0

#### Question 3.1. Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college.

In [20]:
pd.crosstab(df.prestige_1, df.admit, margins=True)

admit         0    1  All
prestige_1               
0.0         243   93  336
1.0          28   33   61
All         271  126  397

calculate oods: (the number of desired divided by the number of non-desired values)
reference a row by the prestige_x or Px value == 1, and then reference the admit vs not-admit column of that row.

Answer:
ODDS of getting in if from prestige_1 school: 33:28
Proability: 33/61 = .5409 = 54.09%
OR CAN CALCULATE ODDS: .5409/(1-.5409) = 1.18 (same as 33/28)

#### Question 3.2. Now calculate the odds of admission if you did not attend a #1 ranked college.

Answer: 

Probability: 93/336 = 0.2768 = 27.68%

**Odds: 93:243 = 31:81**

#### Question 3.3. Calculate the odds ratio.

odds of admitance w/ prestige school : odds of admitance w/o prestige school
Odds ratio (odds of getting accepted from a prestige 1 school vs. getting admitted from not a prestige 1 school):
(33/28)/(93/243) = 3.08 : 1

#### Question 3.4. Write this finding in a sentenance:

In [ ]:

odds of being accepted from a prestige 1 school is 3x greater (308% greater) than being accepted from a different prestige-ranked school

Answer:

#### Question 3.5. Print the cross tab for prestige_4.

#### Question 3.6. Calculate the OR.

#### Question 3.7. Write this finding in a sentence.

Answer:

## Part 4. Analysis

In [15]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
data = df[cols_to_keep].join(prestige_df.ix[:, 'prestige_2':])
print data.head()

   admit    gre   gpa  prestige_2  prestige_3  prestige_4
0      0  380.0  3.61         0.0         1.0         0.0
1      1  660.0  3.67         0.0         1.0         0.0
2      1  800.0  4.00         0.0         0.0         0.0
3      1  640.0  3.19         0.0         0.0         1.0
4      0  520.0  2.93         0.0         0.0         1.0


We're going to add a constant term for our Logistic Regression.  The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [16]:
# manually add the intercept
data['intercept'] = 1.0

#### Question 4.1. Set the covariates to a variable called train_cols.

#### Question 4.2. Fit the model.

#### Question 4.3. Print the summary results.

#### Question 4.4. Calculate the odds ratios of the coeffincients and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params

        conf.columns = ['2.5%', '97.5%', 'OR']

#### Question 4.5. Interpret the OR of Prestige_2.

Answer:

#### Question 4.6. Interpret the OR of GPA.

Answer:

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values.  This will allow us to see how the predicted probability of admission increases/decreases across different variables.  First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa".  This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [ ]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [ ]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max
gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)

print gres
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])

gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)

print gpas
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])

# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))

#### Question 5.1. Recreate the dummy variables.

In [ ]:
# recreate the dummy variables

# keep only what we need for making predictions


#### Question 5.2. Make predictions on the enumerated dataset.

#### Question 5.3. Interpret findings for the last 4 observations.

Answer:

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.